src/wayland/mod.rs

```diff
@@ -72,60 +73,119 @@ impl WaylandImpl {
+        self.size = Some(
+            (|| {
+                let width = NonZeroI32::new(i32::try_from(width).ok()?)?;
+                let height = NonZeroI32::new(i32::try_from(height).ok()?)?;
+                Some((width, height))
+            })()
+            .ok_or(SoftBufferError::SizeOutOfRange { width, height })?,
+        );
+        Ok(())
+    }
+
+    pub fn buffer_mut(&mut self) -> Result<BufferImpl, SoftBufferError> {
+        let (width, height) = self
+            .size
+            .expect("Must set size of surface before calling `buffer_mut()`");
+
+        if let Some((_front, back)) = &mut self.buffers {
+            // Block if back buffer not released yet
             if !back.released() {
-                let mut event_queue = self.display.event_queue.lock().unwrap();
+                let mut event_queue = self.display.event_queue.borrow_mut();
                 while !back.released() {
-                    event_queue.blocking_dispatch(&mut State).unwrap();
+                    event_queue.blocking_dispatch(&mut State).map_err(|err| {
+                        SoftBufferError::PlatformError(
+                            Some("Wayland dispatch failure".to_string()),
+                            Some(Box::new(err)),
+                        )
+                    })?;
                 }
             }
-            back.resize(width, height);
-            (back, front)
+
+            // Resize, if buffer isn't large enough
+            back.resize(width.into(), height.into());
         } else {
             // Allocate front and back buffer
-            (
-                WaylandBuffer::new(&self.display.shm, width, height, &self.display.qh),
-                WaylandBuffer::new(&self.display.shm, width, height, &self.display.qh),
-            )
-        });
-        &self.buffers.as_ref().unwrap().0
+            self.buffers = Some((
+                WaylandBuffer::new(
+                    &self.display.shm,
+                    width.into(),
+                    height.into(),
+                    &self.display.qh,
+                ),
+                WaylandBuffer::new(
+                    &self.display.shm,
+                    width.into(),
+                    height.into(),
+                    &self.display.qh,
+                ),
+            ));
+        };
+
+        Ok(BufferImpl(util::BorrowStack::new(self, |buffer| {
+            Ok(unsafe { buffer.buffers.as_mut().unwrap().1.mapped_mut() })
+        })?))
     }
+}
+
+pub struct BufferImpl<'a>(util::BorrowStack<'a, WaylandImpl, [u32]>);
 
-    pub(super) unsafe fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
-        let _ = self
+impl<'a> BufferImpl<'a> {
+    #[inline]
+    pub fn pixels(&self) -> &[u32] {
+        self.0.member()
+    }
+
+    #[inline]
+    pub fn pixels_mut(&mut self) -> &mut [u32] {
+        self.0.member_mut()
+    }
+
+    pub fn present(self) -> Result<(), SoftBufferError> {
```

Comment:
```suggestion
    pub fn present(mut self) -> Result<(), SoftBufferError> {
```
This should allow remove `RefCell`.